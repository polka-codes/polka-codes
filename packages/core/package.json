{
  "name": "@polka-codes/core",
  "version": "0.8.17",
  "license": "AGPL-3.0",
  "author": "github@polka.codes",
  "type": "module",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "types": "./dist/index.d.ts"
    }
  },
  "files": ["dist"],
  "scripts": {
    "build": "tsup src/index.ts --experimental-dts --format esm --clean"
  },
  "dependencies": {
    "@anthropic-ai/sdk": "0.39.0",
    "lodash": "^4.17.21",
    "openai": "^4.80.0",
    "yaml": "^2.7.1",
    "zod": "^3.24.2"
  }
}
