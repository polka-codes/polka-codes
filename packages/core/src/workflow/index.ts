// generated by polka.codes

export * from './agent'
export * from './builder'
export * from './command'
export * from './handlers'
export * from './dsl'
export * from './steps'
export * from './types'
export * from './workflow'
