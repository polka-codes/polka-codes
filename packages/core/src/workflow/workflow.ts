// generated by polka.codes

import { runStep } from './runStep'
import type { Json, StepSpecHandlerFn, WorkflowContext, WorkflowRunResult, WorkflowSpec } from './types'

export const run = async <TInput extends Record<string, Json>, TOutput extends Record<string, Json>>(
  workflow: WorkflowSpec<TInput, TOutput>,
  context: WorkflowContext,
  handler: StepSpecHandlerFn<any, any>,
  input: TInput,
): Promise<WorkflowRunResult<TOutput>> => {
  const rootStep = handler(workflow.step, handler)
  const result = await runStep(rootStep, input, context, undefined, {})
  switch (result.type) {
    case 'paused':
      return {
        type: 'paused',
        state: result.state,
      }
    case 'error':
      return {
        type: 'error',
        error: result.error,
      }
    case 'success':
      return {
        type: 'success',
        output: result.output as TOutput,
      }
  }
}

export const resume = async <TInput extends Record<string, Json>, TOutput extends Record<string, Json>>(
  workflow: WorkflowSpec<TInput, TOutput>,
  context: WorkflowContext,
  handler: StepSpecHandlerFn<any, any>,
  state: any,
  input: TInput, // This is the new input for the resumed workflow
): Promise<WorkflowRunResult<TOutput>> => {
  const rootStep = handler(workflow.step, handler)
  const result = await runStep(rootStep, input, context, state, {})
  switch (result.type) {
    case 'paused':
      return {
        type: 'paused',
        state: result.state,
      }
    case 'error':
      return {
        type: 'error',
        error: result.error,
      }
    case 'success':
      return {
        type: 'success',
        output: result.output as TOutput,
      }
  }
}
