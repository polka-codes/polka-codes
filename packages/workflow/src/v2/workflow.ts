import type { Logger, ToolRegistry } from '../workflow'

export type ToolHandler<TTools extends ToolRegistry> = {
  [K in keyof TTools]: (input: TTools[K]['input']) => Promise<TTools[K]['output']>
}

export type WorkflowContextV2<TTools extends ToolRegistry> = {
  step: <T>(name: string, fn: () => Promise<T>) => Promise<T>
  logger: Logger
  toolHandler: ToolHandler<TTools>
}

export type WorkflowFnV2<TInput, TOutput, TTools extends ToolRegistry> = (
  input: TInput,
  context: WorkflowContextV2<TTools>,
) => Promise<TOutput>

// Create a default silent logger
const silentLogger: Logger = {
  debug: () => {},
  info: () => {},
  warn: () => {},
  error: () => {},
}

export function createContext<TTools extends ToolRegistry>(
  toolHandler: ToolHandler<TTools>,
  stepFn?: <T>(name: string, fn: () => Promise<T>) => Promise<T>,
  logger: Logger = silentLogger,
): WorkflowContextV2<TTools> {
  if (!stepFn) {
    // simple default step function
    stepFn = async <T>(_name: string, fn: () => Promise<T>) => fn()
  }

  return { step: stepFn, logger, toolHandler }
}

export const makeStepFn = (): (<T>(name: string, fn: () => Promise<T>) => Promise<T>) => {
  const results: Map<string, any> = new Map()

  return async <T>(name: string, fn: () => Promise<T>): Promise<T> => {
    if (results.has(name)) {
      return results.get(name) as T
    }

    const result = await fn()
    results.set(name, result)
    return result
  }
}
