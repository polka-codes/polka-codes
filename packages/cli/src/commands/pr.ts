import { execSync, spawnSync } from 'node:child_process'
import { createService, generateGithubPullRequestDetails } from '@polka-codes/core'
import { Command } from 'commander'
import ora from 'ora'

import { parseOptions } from '../options'

export const prCommand = new Command('pr')
  .description('Create a GitHub pull request')
  .argument('[message]', 'Optional context for the commit message generation')
  .action(async (message, options) => {
    const spinner = ora('Gathering information...').start()

    const { providerConfig } = parseOptions(options)

    const { provider, model, apiKey } = providerConfig.getConfigForCommand('pr') ?? {}

    if (!provider) {
      console.error('Error: No provider specified. Please run "polka-codes config" to configure your AI provider.')
      process.exit(1)
    }

    try {
      // Check if gh CLI is installed
      execSync('gh --version', { stdio: 'ignore' })
    } catch {
      console.error('Error: GitHub CLI (gh) is not installed. Please install it from https://cli.github.com/')
      process.exit(1)
    }

    try {
      // get branch name usnig git
      const branchName = execSync('git rev-parse --abbrev-ref HEAD', { encoding: 'utf-8' }).trim()

      const defaultBranchNames = ['master', 'main', 'develop']

      let defaultBranch: string | undefined

      for (const branchName of defaultBranchNames) {
        if (execSync(`git branch --list ${branchName}`, { encoding: 'utf-8' }).includes(branchName)) {
          defaultBranch = branchName
          break
        }
      }

      if (!defaultBranch) {
        // git remote is slow as it needs to fetch the remote info
        const originInfo = execSync('git remote show origin', { encoding: 'utf-8' })
        defaultBranch = originInfo.match(/HEAD branch: (.*)/)?.[1]
      }

      if (!defaultBranch) {
        console.error('Error: Could not determine default branch name.')
        process.exit(1)
      }

      const commits = execSync(`git --no-pager log --oneline --no-color --no-merges --no-decorate ${defaultBranch}..HEAD`, {
        encoding: 'utf-8',
      })

      const diff = execSync(`git diff --cached -U200 ${defaultBranch}`, { encoding: 'utf-8' })

      const ai = createService(provider, {
        apiKey,
        model,
      })

      spinner.text = 'Generating pull request details...'

      const prDetails = await generateGithubPullRequestDetails(ai, {
        branchName: branchName,
        context: message,
        commitMessages: commits,
        commitDiff: diff,
      })

      spinner.succeed('Pull request details generated')

      spawnSync('gh', ['pr', 'create', '--title', prDetails.response.title.trim(), '--body', prDetails.response.description.trim()], {
        stdio: 'inherit',
      })
    } catch (error) {
      console.error('Error creating pull request:', error)
      process.exit(1)
    }
  })
