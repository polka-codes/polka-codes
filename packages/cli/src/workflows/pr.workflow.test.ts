// generated by polka.codes

import { expect, mock, test } from 'bun:test'
import { ToolResponseType } from '@polka-codes/core'
import type { WorkflowContextV2 } from '@polka-codes/workflow'
import { createContext } from '@polka-codes/workflow'
import type { CliToolRegistry } from '../workflow-tools'
import { prWorkflow } from './pr.workflow'

const createMockContext = () => {
  const toolHandler = {
    executeCommand: mock(async () => ({ exitCode: 0, stdout: '', stderr: '' })),
    createPullRequest: mock(async (details) => details),
    generateText: mock(async () => {
      return [
        {
          role: 'assistant',
          content: JSON.stringify({
            title: 'feat: add authentication feature',
            description:
              'This PR implements the authentication feature for the application.\n\n### Changes\n- Added authenticate function\n- Fixed login bug',
          }),
        },
      ]
    }),
    taskEvent: mock(async () => {}),
    // Add other necessary mocks from CliToolRegistry here
    invokeAgent: mock(async () => ({ output: {}, messages: [] })),
    createCommit: mock(async () => ({ message: '' })),
    printChangeFile: mock(async () => ({ stagedFiles: [], unstagedFiles: [] })),
    confirm: mock(async () => true),
    input: mock(async () => ''),
    select: mock(async () => ''),
    writeToFile: mock(async () => {}),
    readFile: mock(async () => ''),
    invokeTool: mock(async () => ({ type: ToolResponseType.Reply, message: { type: 'text', value: '' } })),
  }

  const run = (input: any) => {
    const context = createContext<CliToolRegistry>(toolHandler as any)
    const testContext: WorkflowContextV2<CliToolRegistry> = {
      ...context,
      step: (_name, fn) => {
        // In tests, we can simplify the step function to execute immediately
        return fn()
      },
    }
    return prWorkflow(input, testContext)
  }

  return {
    toolHandler,
    run,
  }
}

test('should create PR with user context', async () => {
  const { toolHandler, run } = createMockContext()

  // Mock the sequence of executeCommand calls
  toolHandler.executeCommand
    .mockResolvedValueOnce({ exitCode: 0, stdout: 'gh version 2.40.0', stderr: '' }) // gh --version
    .mockResolvedValueOnce({ exitCode: 0, stdout: 'feature/add-auth\n', stderr: '' }) // git rev-parse
    .mockResolvedValueOnce({ exitCode: 1, stdout: '', stderr: '' }) // git show-ref master
    .mockResolvedValueOnce({ exitCode: 0, stdout: '', stderr: '' }) // git show-ref main
    .mockResolvedValueOnce({
      exitCode: 0,
      stdout: 'abc1234 feat: add authentication module\ndef5678 fix: resolve login bug\n',
      stderr: '',
    }) // git log
    .mockResolvedValueOnce({
      exitCode: 0,
      stdout: `--- a/src/auth.ts
+++ b/src/auth.ts
@@ -1,3 +1,5 @@
+export function authenticate() {
+  return true;
+}
 export function login() {}`,
      stderr: '',
    }) // git diff

  const result = await run({ context: 'Implementing new authentication feature' })

  expect(result).toEqual({
    title: 'feat: add authentication feature',
    description:
      'This PR implements the authentication feature for the application.\n\n### Changes\n- Added authenticate function\n- Fixed login bug',
  })

  expect(toolHandler.executeCommand).toHaveBeenCalledTimes(6)
  expect(toolHandler.generateText).toHaveBeenCalledTimes(1)
  expect(toolHandler.createPullRequest).toHaveBeenCalledTimes(1)
})

test('should fail when GitHub CLI is not installed', async () => {
  const { toolHandler, run } = createMockContext()

  toolHandler.executeCommand.mockResolvedValueOnce({
    exitCode: 1,
    stdout: '',
    stderr: 'command not found: gh',
  })

  await expect(run({})).rejects.toThrow('GitHub CLI (gh) is not installed. Please install it from https://cli.github.com/')
})

test('should fail when unable to determine default branch', async () => {
  const { toolHandler, run } = createMockContext()

  toolHandler.executeCommand
    .mockResolvedValueOnce({ exitCode: 0, stdout: 'gh version 2.40.0', stderr: '' }) // gh --version
    .mockResolvedValueOnce({ exitCode: 0, stdout: 'feature/my-branch\n', stderr: '' }) // git rev-parse
    .mockResolvedValueOnce({ exitCode: 1, stdout: '', stderr: 'not a git repository' }) // gh repo view
    .mockResolvedValueOnce({ exitCode: 1, stdout: '', stderr: '' }) // git show-ref master
    .mockResolvedValueOnce({ exitCode: 1, stdout: '', stderr: '' }) // git show-ref main
    .mockResolvedValueOnce({ exitCode: 1, stdout: '', stderr: '' }) // git show-ref develop
    .mockResolvedValueOnce({ exitCode: 1, stdout: '', stderr: 'fatal: No remote named origin' }) // git remote show origin

  await expect(run({})).rejects.toThrow('Could not determine default branch name.')
})
