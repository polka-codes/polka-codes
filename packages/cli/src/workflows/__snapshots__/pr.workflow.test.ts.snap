// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`should create PR with user context 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--version",
      ],
      "command": "gh",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR with user context 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "rev-parse",
        "--abbrev-ref",
        "HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR with user context 3`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "show-ref",
        "--verify",
        "--quiet",
        "refs/heads/master",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR with user context 4`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "show-ref",
        "--verify",
        "--quiet",
        "refs/heads/main",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR with user context 5`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--no-pager",
        "log",
        "--oneline",
        "--no-color",
        "--no-merges",
        "--no-decorate",
        "main..HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR with user context 6`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--no-pager",
        "diff",
        "main..HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR with user context 7`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "kind": "StartTask",
      "systemPrompt": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
    },
    "tool": "taskEvent",
    "type": "tool",
  },
}
`;

exports[`should create PR with user context 8`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "kind": "StartRequest",
      "userMessage": [
        {
          "content": 
"<branch_name>feature/add-auth</branch_name>
<commit_messages>abc1234 feat: add authentication module
def5678 fix: resolve login bug
</commit_messages>
<diff>--- a/src/auth.ts
+++ b/src/auth.ts
@@ -1,3 +1,5 @@
+export function authenticate() {
+  return true;
+}
 export function login() {}
</diff>
<user_context>Implementing new authentication feature</user_context>"
,
          "role": "user",
        },
      ],
    },
    "tool": "taskEvent",
    "type": "tool",
  },
}
`;

exports[`should create PR with user context 9`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "messages": [
        {
          "content": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
          "role": "system",
        },
        {
          "content": 
"<branch_name>feature/add-auth</branch_name>
<commit_messages>abc1234 feat: add authentication module
def5678 fix: resolve login bug
</commit_messages>
<diff>--- a/src/auth.ts
+++ b/src/auth.ts
@@ -1,3 +1,5 @@
+export function authenticate() {
+  return true;
+}
 export function login() {}
</diff>
<user_context>Implementing new authentication feature</user_context>"
,
          "role": "user",
        },
      ],
      "tools": {},
    },
    "tool": "generateText",
    "type": "tool",
  },
}
`;

exports[`should create PR without context 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--version",
      ],
      "command": "gh",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR without context 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "rev-parse",
        "--abbrev-ref",
        "HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR without context 3`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "show-ref",
        "--verify",
        "--quiet",
        "refs/heads/master",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR without context 4`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "show-ref",
        "--verify",
        "--quiet",
        "refs/heads/main",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR without context 5`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--no-pager",
        "log",
        "--oneline",
        "--no-color",
        "--no-merges",
        "--no-decorate",
        "main..HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR without context 6`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--no-pager",
        "diff",
        "main..HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should create PR without context 7`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "kind": "StartTask",
      "systemPrompt": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
    },
    "tool": "taskEvent",
    "type": "tool",
  },
}
`;

exports[`should create PR without context 8`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "kind": "StartRequest",
      "userMessage": [
        {
          "content": 
"<branch_name>feature/new-feature</branch_name>
<commit_messages>xyz9876 chore: update dependencies
</commit_messages>
<diff>--- a/package.json
+++ b/package.json
@@ -10,7 +10,7 @@
-    "lodash": "^4.17.20"
+    "lodash": "^4.17.21"
</diff>"
,
          "role": "user",
        },
      ],
    },
    "tool": "taskEvent",
    "type": "tool",
  },
}
`;

exports[`should create PR without context 9`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "messages": [
        {
          "content": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
          "role": "system",
        },
        {
          "content": 
"<branch_name>feature/new-feature</branch_name>
<commit_messages>xyz9876 chore: update dependencies
</commit_messages>
<diff>--- a/package.json
+++ b/package.json
@@ -10,7 +10,7 @@
-    "lodash": "^4.17.20"
+    "lodash": "^4.17.21"
</diff>"
,
          "role": "user",
        },
      ],
      "tools": {},
    },
    "tool": "generateText",
    "type": "tool",
  },
}
`;

exports[`should fail when GitHub CLI is not installed 1`] = `
{
  "error": "GitHub CLI (gh) is not installed. Please install it from https://cli.github.com/",
  "status": "failed",
}
`;

exports[`should fail when unable to determine default branch 1`] = `
{
  "error": "Could not determine default branch name.",
  "status": "failed",
}
`;

exports[`should handle branch with special characters 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "show-ref",
        "--verify",
        "--quiet",
        "refs/heads/master",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should handle branch with special characters 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "kind": "StartTask",
      "systemPrompt": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
    },
    "tool": "taskEvent",
    "type": "tool",
  },
}
`;

exports[`should handle branch with special characters 3`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "messages": [
        {
          "content": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
          "role": "system",
        },
        {
          "content": 
"<branch_name>feature/fix-bug-#123</branch_name>
<commit_messages>aaa1111 fix: resolve issue #123
</commit_messages>
<diff>--- a/src/bugfix.ts
+++ b/src/bugfix.ts
@@ -5,7 +5,7 @@
-  return false;
+  return true;
</diff>"
,
          "role": "user",
        },
      ],
      "tools": {},
    },
    "tool": "generateText",
    "type": "tool",
  },
}
`;

exports[`should handle no commits between branches 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--no-pager",
        "diff",
        "master..HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should handle no commits between branches 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "kind": "StartTask",
      "systemPrompt": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
    },
    "tool": "taskEvent",
    "type": "tool",
  },
}
`;

exports[`should handle no commits between branches 3`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "messages": [
        {
          "content": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
          "role": "system",
        },
        {
          "content": 
"<branch_name>feature/empty-branch</branch_name>
<commit_messages></commit_messages>
<diff>--- a/README.md
+++ b/README.md
@@ -1 +1,2 @@
 # Project
+Updated readme
</diff>"
,
          "role": "user",
        },
      ],
      "tools": {},
    },
    "tool": "generateText",
    "type": "tool",
  },
}
`;

exports[`should handle empty diff 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "kind": "StartTask",
      "systemPrompt": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
    },
    "tool": "taskEvent",
    "type": "tool",
  },
}
`;

exports[`should handle empty diff 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "messages": [
        {
          "content": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
          "role": "system",
        },
        {
          "content": 
"<branch_name>feature/empty-diff</branch_name>
<commit_messages>bbb2222 chore: trigger CI
</commit_messages>
<diff></diff>"
,
          "role": "user",
        },
      ],
      "tools": {},
    },
    "tool": "generateText",
    "type": "tool",
  },
}
`;

exports[`should handle multiple commits with various messages 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--no-pager",
        "diff",
        "main..HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should handle multiple commits with various messages 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "kind": "StartTask",
      "systemPrompt": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
    },
    "tool": "taskEvent",
    "type": "tool",
  },
}
`;

exports[`should handle multiple commits with various messages 3`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "messages": [
        {
          "content": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
          "role": "system",
        },
        {
          "content": 
"<branch_name>feature/multi-commit</branch_name>
<commit_messages>abc1234 feat: add user profile page
def5678 fix: resolve navigation issue
ghi9012 test: add unit tests
jkl3456 docs: update API documentation
mno7890 refactor: clean up code
</commit_messages>
<diff>--- a/src/profile.ts
+++ b/src/profile.ts
@@ -0,0 +1,10 @@
+export function getUserProfile() {
+  return { name: 'User' };
+}
--- a/src/nav.ts
+++ b/src/nav.ts
@@ -5,7 +5,7 @@
-  router.push('/wrong');
+  router.push('/correct');
--- a/docs/API.md
+++ b/docs/API.md
@@ -1 +1,5 @@
 # API
+
+## Endpoints
+- GET /profile
</diff>"
,
          "role": "user",
        },
      ],
      "tools": {},
    },
    "tool": "generateText",
    "type": "tool",
  },
}
`;

exports[`should use develop as default branch 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--no-pager",
        "log",
        "--oneline",
        "--no-color",
        "--no-merges",
        "--no-decorate",
        "develop..HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should use develop as default branch 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "messages": [
        {
          "content": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
          "role": "system",
        },
        {
          "content": 
"<branch_name>feature/new-feature</branch_name>
<commit_messages>ccc3333 feat: implement feature
</commit_messages>
<diff>--- a/src/feature.ts
+++ b/src/feature.ts
@@ -1 +1,3 @@
+export function feature() {}
</diff>"
,
          "role": "user",
        },
      ],
      "tools": {},
    },
    "tool": "generateText",
    "type": "tool",
  },
}
`;

exports[`should use gh repo view to determine default branch 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--no-pager",
        "log",
        "--oneline",
        "--no-color",
        "--no-merges",
        "--no-decorate",
        "trunk..HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should use gh repo view to determine default branch 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "messages": [
        {
          "content": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
          "role": "system",
        },
        {
          "content": 
"<branch_name>feature/custom-branch</branch_name>
<commit_messages>ddd4444 refactor: code cleanup
</commit_messages>
<diff>--- a/src/cleanup.ts
+++ b/src/cleanup.ts
@@ -1,5 +1,3 @@
-// old code
-// more old code
 // clean code
</diff>"
,
          "role": "user",
        },
      ],
      "tools": {},
    },
    "tool": "generateText",
    "type": "tool",
  },
}
`;

exports[`should use git remote show origin to determine default branch 1`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "args": [
        "--no-pager",
        "log",
        "--oneline",
        "--no-color",
        "--no-merges",
        "--no-decorate",
        "production..HEAD",
      ],
      "command": "git",
    },
    "tool": "executeCommand",
    "type": "tool",
  },
}
`;

exports[`should use git remote show origin to determine default branch 2`] = `
{
  "next": [Function: AsyncFunction],
  "status": "pending",
  "throw": [Function: AsyncFunction],
  "tool": {
    "input": {
      "messages": [
        {
          "content": 
"
You are an expert at creating pull requests.
Based on the provided branch name, commit messages, and diff, generate a title and description for the pull request.

Respond with a JSON object containing the title and description.
Example format:
\`\`\`json
{
  "title": "feat: add new feature",
  "description": "This pull request adds a new feature that does...\\n\\n### Changes\\n- ..."
}
\`\`\`
"
,
          "role": "system",
        },
        {
          "content": 
"<branch_name>feature/test-branch</branch_name>
<commit_messages>eee5555 build: update config
</commit_messages>
<diff>--- a/config.yml
+++ b/config.yml
@@ -1 +1,2 @@
 version: 1.0
+env: production
</diff>"
,
          "role": "user",
        },
      ],
      "tools": {},
    },
    "tool": "generateText",
    "type": "tool",
  },
}
`;
